# Peer-graded Assignment: Build a Regression Model in Keras

In [4]:
# Import libraries
import keras
from keras.models import Sequential
from keras.layers import Dense
import pandas as pd
import numpy as np
import requests
from io import StringIO
from sklearn.model_selection import train_test_split
import warnings

warnings.simplefilter('ignore', FutureWarning)

In [5]:
# Load dataset
concrete_data = pd.read_csv(StringIO(requests.get(
    'https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv'
).text))

# Defining predictors and target
predictors = concrete_data.drop(columns=['Strength'])
target = concrete_data['Strength']

# Split data into training (70%) and testing (30%) sets
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

## A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the 
train_test_split
helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

Submit your Jupyter Notebook with your code and comments.

In [6]:
# Number of predictors
n_cols = predictors.shape[1]

# Define the regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# Create a list to store the mean squared errors
mse_list = []

# Repeat the process 50 times
for i in range(50):
    # Build and train the model
    model = regression_model()
    model.fit(X_test, y_test, epochs=50, verbose=0) 

    # Evaluate the model on the validation set
    test_mse = model.evaluate(X_test, y_test, verbose=0)
    mse_list.append(test_mse)

# Compute mean and standard deviation of MSE
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f'Mean MSE: {mean_mse:.4f}')
print(f'Standard Deviation MSE: {std_mse:.4f}')

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object 

Mean MSE: 905.4717
Standard Deviation MSE: 1108.7471


## B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?


In [7]:
# Normalize the predictors
predictors = (predictors - predictors.mean()) / predictors.std()

# Split data into training (70%) and testing (30%) sets
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

# Number of predictors
n_cols = predictors.shape[1]

# Define the regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# List to store mean squared errors
mse_list = []

# Repeat 50 times
for i in range(50):
    # Build and train model
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    
    # Evaluate the model on the validation set
    test_mse = model.evaluate(X_test, y_test, verbose=0)
    mse_list.append(test_mse)

# Compute mean and standard deviation of MSE
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"Mean MSE: {mean_mse:.4f}")
print(f"Standard Deviation MSE: {std_mse:.4f}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object 

Mean MSE: 342.9756
Standard Deviation MSE: 103.2880


</p>Mean MSE dropped from 413.1677 to 371.4195, meaning the model predicts better on average.
</p>Standard Deviation MSE decreased drastically from 506.8472 to 91.0009, indicating that the model's performance is more consistent across different runs.

## C. Increase the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [10]:
# Normalize the predictors
predictors = (predictors - predictors.mean()) / predictors.std()

# Split data into training (70%) and testing (30%) sets
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

# Number of predictors
n_cols = predictors.shape[1]

# Define the regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# List to store mean squared errors
mse_list = []

# Repeat 50 times
for i in range(50):
    # Build and train model
    model = regression_model()
    model.fit(X_train, y_train, epochs=100, verbose=0)
    
    # Evaluate the model on the validation set
    test_mse = model.evaluate(X_test, y_test, verbose=0)
    mse_list.append(test_mse)

# Compute mean and standard deviation of MSE
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"Mean MSE: {mean_mse:.4f}")
print(f"Standard Deviation MSE: {std_mse:.4f}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object 

Mean MSE: 159.5040
Standard Deviation MSE: 20.5944


</p>Mean MSE dropped significantly (from 371.4195 to 181.7803)
→ The model is now making much more accurate predictions.
</p>Standard Deviation decreased further (from 91.0009 to 24.7290)
→ The model’s performance is even more consistent across different runs.
</p>More epochs helped significantly → The model had more time to learn and generalize better.

## D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [9]:
# Normalize the predictors
predictors = (predictors - predictors.mean()) / predictors.std()

# Split data into training (70%) and testing (30%) sets
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

# Number of predictors
n_cols = predictors.shape[1]

# Define the regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# List to store mean squared errors
mse_list = []

# Repeat 50 times
for i in range(50):
    # Build and train model
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    
    # Evaluate the model on the validation set
    test_mse = model.evaluate(X_test, y_test, verbose=0)
    mse_list.append(test_mse)

# Compute mean and standard deviation of MSE
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"Mean MSE: {mean_mse:.4f}")
print(f"Standard Deviation MSE: {std_mse:.4f}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object 

Mean MSE: 125.8408
Standard Deviation MSE: 11.9127


</p>Adding more hidden layers (from 1 to 3) improved accuracy
</p>Comparing Part B with Part D, we can conclude that:
</p>Mean MSE dropped from 371.4195 to 138.3501. A deeper network captures more complex patterns in the data.
</p>The Standard Deviation remains low (24.2872), as in part C. The model’s results are still stable across different runs.
</p>This happened because more hidden layers allow the network to extract hierarchical features from the data.
This improves the model’s ability to generalize, even with the same number of epochs.

## Testing by myself: C + D (100 epochs, 3 layers, normalized)

In [11]:
# Normalize the predictors
predictors = (predictors - predictors.mean()) / predictors.std()

# Split data into training (70%) and testing (30%) sets
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

# Number of predictors
n_cols = predictors.shape[1]

# Define the regression model
def regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# List to store mean squared errors
mse_list = []

# Repeat 50 times
for i in range(50):
    # Build and train model
    model = regression_model()
    model.fit(X_train, y_train, epochs=100, verbose=0)
    
    # Evaluate the model on the validation set
    test_mse = model.evaluate(X_test, y_test, verbose=0)
    mse_list.append(test_mse)

# Compute mean and standard deviation of MSE
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

print(f"Mean MSE: {mean_mse:.4f}")
print(f"Standard Deviation MSE: {std_mse:.4f}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object 

Mean MSE: 83.7522
Standard Deviation MSE: 21.6973


Final Comparison of Results:
| Step  | Mean MSE  | Standard Deviation MSE  |
|-------|-----------|-------------------------|
| **A** (Baseline, 50 epochs, 1 layer, unnormalized) | **905.4717** | **1108.7471** |
| **B** (50 epochs, 1 layer, normalized) | **342.9756** | **103.2880** |
| **C** (100 epochs, 1 layer, normalized) | **159.5040** | **20.5944** |
| **D** (50 epochs, 3 layers, normalized) | **125.8408** | **11.9127** |
| **C + D** (100 epochs, 3 layers, normalized) | **83.7522** | **21.6973** |


</p>✅ Mean MSE keeps decreasing → Predictions are getting more accurate.
</p>✅ Standard Deviation remains low → The model is still stable, though slightly higher than in Part D.
</p>✅ More epochs + deeper network → The combination of increased training time and better feature extraction led to the best performance yet!